File: src/shared/routerUtils.ml
open Router.Route

let requireSubsonicCredentials { context = { store }; location; _ } =
  let state = store |. Zustand.getState in
  (match state.auth.credentials with
  | None ->
      Some
        Router.Redirect.(
          options ~replace:true
            ~search:[%mel.obj { next = location.pathname ^ location.searchStr }]
            ~_to:"/login" ()
          |> make)
  | Some _ -> None)
  |> Js.Promise.resolve

File: src/shared/routerUtils.ts
import type { ParsedLocation } from '@tanstack/react-router';
import { redirect } from '@tanstack/react-router';

import type { Store } from '../store/types';

export function requireSubsonicCredentials({
  context,
  location,
}: {
  context: { store: Store };
  location: ParsedLocation;
}) {
  const state = context.store.getState();

  return state.auth.credentials
    ? undefined
    : redirect({
        replace: true,
        search: { next: location.pathname + location.searchStr },
        to: '/login',
      });
}
